src/objective/regression_objective.hpp

```diff
@@ -161,6 +161,73 @@ class RegressionL1loss: public RegressionL2loss {
   double eta_;
 };
 
+/*!
+* \brief SMAPE regression loss
+*/
+class RegressionSMAPEloss: public RegressionL2loss {
+public:
+  explicit RegressionSMAPEloss(const ObjectiveConfig& config): RegressionL2loss(config) { }
+
+  explicit RegressionSMAPEloss(const std::vector<std::string>& strs): RegressionL2loss(strs) { }
+
+  ~RegressionSMAPEloss() {}
+
+  void GetGradients(const double* score, score_t* gradients,
+                    score_t* hessians) const override {
+    if (weights_ == nullptr) {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        const double diff = score[i] - label_[i];
+        const double sumabs = std::fabs(score[i]) + std::fabs(label_[i]);
+        if (diff == 0) {
+          gradients[i] = 0.0f;
+          hessians[i] = 2.0f / sumabs;
+        } else {
+          const double absdiff = std::fabs(diff);
+          const double sumabs2 = sumabs * sumabs;
+          const double sumabs3 = sumabs2 * sumabs;
+          const double t1 = (2.0f / sumabs3 - (2.0f * (score[i] == 0)) / (label_[i] * label_[i]));
+          const double t2 = 2.0f  * sgn(label_[i]) * sgn(diff) / sumabs2;
+          gradients[i] = (sgn(diff) * sumabs - sgn(score[i]) * absdiff) / sumabs2;
+          hessians[i] = absdiff * t1 - t2;
+        }
+      }
+    } else {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        const double diff = score[i] - label_[i];
+        const double sumabs = std::fabs(score[i]) + std::fabs(label_[i]);
+        if (diff == 0) {
+          gradients[i] = 0.0f;
+          hessians[i] = 2.0f * weights_[i] / sumabs;
+        } else {
+          const double absdiff = std::fabs(diff);
+          const double sumabs2 = sumabs * sumabs;
+          const double sumabs3 = sumabs2 * sumabs;
+          const double t1 = (2.0f / sumabs3 - (2.0f * (score[i] == 0)) / (label_[i] * label_[i]));
+          const double t2 = 2.0f  * sgn(label_[i]) * sgn(diff) / sumabs2;
+          gradients[i] = ((sgn(diff) * sumabs - sgn(score[i]) * absdiff)) * weights_[i] / sumabs2;
```

Comment:
you can use `std::copysign` .

Reply:
```
sgn(a) * b != std::copysign(b, a) // when a = 0
```

Should I still replace sgn with copysign? Perhaps there is a better place to keep copysign so that it can be used in other places as well. A number of places in code can make use of this function rather than having branches.

Reply:
copysign is more efficient, it will avoid the if...else check.

Reply:
I meant that the function I wrote relies on the fact that `sgn(0) = 0`. This is not the case with copysign. I am also unsure about how copysign will avoid if-else. Can you please elaborate?

Reply:
you can see the compiled Assemble code, copysign will use the cpu instructions to get sign directly

Reply:
i think copysign is safe for 0 . you can test it

Reply:
```
  std::cout << std::copysign(1, 0) << std::endl;
```

The above statement print 1 and not 0

Reply:
As far as branching in `sgn` is concerned, the version I wrote does not have any branches https://godbolt.org/g/AUnpfi

Reply:
okay, great